src/reconcile/rec_visibility.c

```diff
@@ -872,26 +981,75 @@ __rec_fill_tw_from_upd_select(
      * that the value is visible to any timestamp/transaction id ahead of it.
      */
     if (upd->type == WT_UPDATE_TOMBSTONE) {
-        WT_TIME_WINDOW_SET_STOP(select_tw, upd, upd->prepare_state);
+        WT_TIME_WINDOW_SET_STOP(select_tw, upd, write_prepare);
         tombstone = upd_select->tombstone = upd;
 
-        /* Find the update this tombstone applies to. */
-        if (!__wt_txn_upd_visible_all(session, upd)) {
-            while (upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED)
-                upd = upd->next;
+        /*
+         * Find the update this tombstone applies to.
+         *
+         * We need to find the full update if the prepared tombstone is rolled back. We resolve
+         * prepare updates recursively from the oldest to the newest. If we write a prepared
+         * tombstone, we may see a prepared update that is rolled back from the same transaction.
+         * Ensure we also write it to the disk in this case.
+         */
+        if (write_prepare || !__wt_txn_upd_visible_all(session, upd)) {
+            for (; upd->next != NULL && upd->next->txnid == WT_TXN_ABORTED; upd = upd->next) {
```

Comment:
The read for upd->next->txnid == WT_TXN_ABORTED needs to use a WT_ACQUIRE_READ

Reply:
not really, only it is related to prepared

Reply:
Here it is needed because we read the upd_saved_txnid. We need to do acquired read if we also read upd_saved_txnid.